app/src/main/java/com/bogdan/codeforceswatcher/features/actions/ActionActivity.kt

```diff
@@ -1,5 +1,105 @@
 import androidx.appcompat.app.AppCompatActivity
+import com.bogdan.codeforceswatcher.R
+import com.bogdan.codeforceswatcher.features.actions.models.CFAction
+import com.bogdan.codeforceswatcher.store
+import kotlinx.android.synthetic.main.activity_action.*
 
-class ActionActivity : AppCompatActivity()
\ No newline at end of file
+class ActionActivity : AppCompatActivity() {
+
+    private var cfAction: CFAction? = null
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_action)
+
+        setSupportActionBar(toolbar)
+        supportActionBar?.setDisplayHomeAsUpEnabled(true)
+        supportActionBar?.setDisplayShowHomeEnabled(true)
+
+        val commentId = intent.getLongExtra(COMMENT_ID, -1)
+        cfAction = store.state.actions.actions.find { it.comment?.id == commentId }
+
+        initViews()
+    }
+
+    private fun initViews() {
+        val pageTitle = cfAction?.blogEntry?.title.orEmpty()
+        val link = getString(R.string.comment_url, cfAction?.blogEntry?.id, cfAction?.comment?.id)
+        tvPageTitle.text = pageTitle
+        btnShare.setOnClickListener { share(link, pageTitle) }
+        setupWebView(link)
+    }
+
+    private fun share(link: String, pageTitle: String) {
+        val share = Intent(Intent.ACTION_SEND)
+        share.type = "text/plain"
+        val shareText = "$pageTitle - $link\n\n${getString(R.string.shared_through_cw)}"
+        share.putExtra(Intent.EXTRA_TEXT, shareText)
+        startActivity(Intent.createChooser(share, getString(R.string.share_with_friends)))
+    }
+
+    @SuppressLint("SetJavaScriptEnabled")
+    private fun setupWebView(url: String) {
+        webView.loadUrl(url)
+
+        webView.settings.javaScriptEnabled = true
+        webView.settings.layoutAlgorithm = WebSettings.LayoutAlgorithm.SINGLE_COLUMN
+        webView.settings.useWideViewPort = true
+        webView.settings.builtInZoomControls = true
+        webView.settings.displayZoomControls = false
+
+        webView.webViewClient = object : WebViewClient() {
+
+            override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
+                webView.loadUrl(url)
+                return true
+            }
+
+            override fun onPageFinished(view: WebView, url: String) {
+                progressBarHorizontal?.visibility = View.GONE
+                webView?.visibility = View.VISIBLE
+            }
+        }
+    }
+
+    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
+        when (keyCode) {
+            KEYCODE_BACK -> {
+                if (webView.canGoBack()) {
+                    webView.goBack()
+                } else {
+                    finish()
+                }
+                return true
+            }
+        }
+        return super.onKeyDown(keyCode, event)
+    }
+
+    override fun onSupportNavigateUp(): Boolean {
+        onBackPressed()
```

Comment:
Hm, why are you handling `back` in 2 places?

Reply:
You say about it second place? `override fun onKeyDown`

Reply:
@bogdan-evtushenko yes